### ISSUE TITLE
[vscode] support CustomEditor API

### ISSUE THREAD

**Issue by tolusha** [eclipse-theia/theia#6636]
### Description

It can be tested with @theia/vscode-builtin-image-preview@latest.

Allows to preview images of such formats:

- `*.jpg`, `*.jpe`, `*.jpeg`
- `*.png`
- `*.bmp`
- `*.gif`
- `*.ico`
- `*.tga`
- `*.webp`

```
root ERROR [hosted-plugin: 5438] Unsupported activation events: onWebviewEditor:imagePreview.previewEditor, please open an issue: https://github.com/eclipse-theia/theia/issues/new
root ERROR [hosted-plugin: 5438] vscode.image-preview extension will be activated eagerly.
root INFO [hosted-plugin: 5438] PLUGIN_HOST(5438): PluginManagerExtImpl/loadPlugin(/home/tolusha/projects/theia-ide/theia/plugins/dir_vscode-builtin-image-preview-0.3.0-next.db32534bf1.tgz/dist/extension)
root ERROR [hosted-plugin: 5438] Error on activation of image-preview TypeError: s.window.registerWebviewEditorProvider is not a function
    at t.activate (/home/tolusha/projects/theia-ide/theia/plugins/dir_vscode-builtin-image-preview-0.3.0-next.db32534bf1.tgz/dist/extension.js:1:7770)
    at PluginManagerExtImpl.<anonymous> (/home/tolusha/projects/theia-ide/theia/packages/plugin-ext/lib/plugin/plugin-manager.js:460:87)
    at step (/home/tolusha/projects/theia-ide/theia/packages/plugin-ext/lib/plugin/plugin-manager.js:47:23)
    at Object.next (/home/tolusha/projects/theia-ide/theia/packages/plugin-ext/lib/plugin/plugin-manager.js:28:53)
    at fulfilled (/home/tolusha/projects/theia-ide/theia/packages/plugin-ext/lib/plugin/plugin-manager.js:19:58)
```

**Comment by akosyakov** [comment]
VS Code is working on new API: Webview Editor. But it is very raw, unstable and not completely implemented. I would prefer to wait till it is stabilized and complete, and only after align with it. Before it we have to find a version of extension which does not use `registerWebviewEditorProvider`


> Unsupported activation events: onWebviewEditor:imagePreview.previewEditor, please open an issue: https://github.com/eclipse-theia/theia/issues/new

That sounds bogus, this activation event is already supported, we should list it to avoid premature activations.

**Comment by akosyakov** [comment]
> Before it we have to find a version of extension which does not use registerWebviewEditorProvider

Ah, this extension was added to prototype webview editors. So there is not version which we can use.

**Comment by vince-fugnitto** [comment]
Adding a link to the proposed API (at the time of writing this comment it looks to be planned for their February 2020 release): https://github.com/microsoft/vscode/issues/77131#issuecomment-567743066

I'm wondering if we should stub the API for the moment, until the proposed API has been approved.

**Comment by akosyakov** [comment]
@vince-fugnitto would it make an extension work? I doubt

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto would it make an extension work? I doubt

I doubt it too, I was wondering if we should stub only so that we can add such `builtins` (latest) without the worry it may have some side-effects.

**Comment by akosyakov** [comment]
@vince-fugnitto not sure what you mean by side effects, if a built-in does not work we should not add it without fixing

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto not sure what you mean by side effects, if a built-in does not work we should not add it without fixing

I may just be overthinking it but I was thinking that perhaps a builtin might work partially (everything besides `WebviewEditor` functionality) and may be useful to add (external extensions may use or refer to such builtins). I thought maybe stubbing those areas until the proposed API is approved might be a workaround.

**Comment by akosyakov** [comment]
We should try for sure and see, but i don't want to port each bug and experiment from VS Code to Theia just to support 1-2 extensions each month. Don't think it is an effort well spent.

But generally yes we should try which latest builtins works and migrate to them and if not collect bugs and reevaluate after each release whether we should take actions on already on them.

**Comment by vince-fugnitto** [comment]
> We should try for sure and see, but i don't want to port each bug and experiment from VS Code to Theia just to support 1-2 extensions each month. Don't think it is an effort well spent.
> 
> But generally yes we should try which latest builtins work and migrate to them and if not collect bugs and reevaluate after each release whether we should take actions on already on them.

Sounds good! I was likely just thinking out loud 🙈

**Comment by akosyakov** [comment]
Not urgent yet, but the python extension already has a mode in which it makes use of custom editors.

**Comment by akosyakov** [comment]
This API was finalised in VS Code: https://code.visualstudio.com/api/extension-guides/custom-editors

**Comment by cbos** [comment]
@akosyakov @vince-fugnitto 
Any idea when this will be implemented in Theia, are the plans for that already?
Markdown and python plugin face issues now.
We faced the same issue as well with another custom plugin.

**Comment by akosyakov** [comment]
We have not planed anything yet. But this issue indeed become important. The current goal is to get proper VS Code fs API support and simplify integration of new Monaco editor. That's enough work till August, so maybe in Autumn.

**Comment by azatsarynnyy** [comment]
Python extension [uses `CustomEditor API`](https://github.com/microsoft/vscode-python/blob/e7136b1448d27e9a607ca0d8242d4272b48aabdb/src/client/datascience/interactive-ipynb/nativeEditorProvider.ts#L35) for rendering Jupyter Notebooks.

Some related references:
- https://code.visualstudio.com/api/extension-guides/custom-editors
- https://github.com/microsoft/vscode/blob/cd3a3ea6faa9d83dae1b706cfe1aaa2dbcaf3411/src/vs/vscode.d.ts#L8067-L8102

We'll start working on it most likely by the end of June.

**Comment by azatsarynnyy** [comment]
fyi: we have to postpone this work for some time

**Comment by amiramw** [comment]
@azatsarynnyy is someone working on this item or we can take it?

**Comment by azatsarynnyy** [comment]
@amiramw you can take it as we won't be able to look at it for the next 1-2 month

**Comment by marcdumais-work** [comment]
Thanks @amiramw for picking-up this one - I hope we're able to include your merged PR in next month's release.

**Comment by hansbank** [comment]
@amiramw , will your work also include the implementation of registerCustomEditorProvider ?

**Comment by EstherPerelman** [comment]
> @amiramw , will your work also include the implementation of registerCustomEditorProvider ?

Yes, It will

**Comment by hansbank** [comment]
Hi @azatsarynnyy , I see this topic has been removed from 1.9.0 milestone. Is there a new milestone set when we can expect this? Is there probably already a version available we can try?
Thank you

**Comment by amiramw** [comment]
@hansbank we are working on it. Hopefully @danarad05 can create initial PR for review next week.

**Comment by danarad05** [comment]
@tolusha @akosyakov @vince-fugnitto @hansbank @azatsarynnyy @amiramw @marcdumais-work @benoitf @cbos @hallvard

FYI, created PR for this issue. Anyone interested in reviewing is welcome.

**Comment by hansbank** [comment]
Hi @danarad05 ,
We have tried the custom editor api for our own editor, and that works as expected. Thank you!
When can we expect this to be merged into theia master?

**Comment by danarad05** [comment]
Hi @hansbank 
Thanks for testing it out. If you are a reviewer then you can review it. 
Currently no one has yet to review this PR so I cannot estimate when it will be merged.
Thanks.
